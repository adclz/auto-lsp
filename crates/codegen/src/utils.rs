/*
This file is part of auto-lsp.
Copyright (C) 2025 CLAUZEL Adrien

auto-lsp is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>
*/

pub(crate) fn sanitize_string(string: &str) -> String {
    if let Some(v) = PUNCTUATION.get(string) {
        v.to_string()
    } else if let Some(v) = RUST_KEYWORDS.get(string) {
        v.to_string()
    } else {
        let mut result = String::new();
        for c in string.chars() {
            if c == '_' {
                result.push(c);
            } else if let Some(v) = PUNCTUATION.get(&c.to_string()) {
                result.push_str(v);
            } else {
                result.push(c);
            }
        }
        result.replace(" ", "_")
    }
}

pub(crate) fn sanitize_string_to_pascal(string: &str) -> String {
    if let Some(v) = PUNCTUATION.get(string) {
        v.to_string()
    } else if let Some(v) = RUST_KEYWORDS.get(string) {
        v.to_string()
    } else {
        let mut result = String::new();
        for c in string.chars() {
            if c == '_' {
                result.push(c);
            } else if let Some(v) = PUNCTUATION.get(&c.to_string()) {
                result.push_str(v);
            } else {
                result.push(c);
            }
        }
        result
            .replace(" ", "_")
            .split('_')
            .map(|word| {
                let mut chars = word.chars();
                match chars.next() {
                    Some(first) => format!(
                        "{}{}",
                        first.to_ascii_uppercase(),
                        chars.collect::<String>()
                    ),
                    None => String::new(),
                }
            })
            .collect()
    }
}

pub static PUNCTUATION: phf::Map<&'static str, &'static str> = phf::phf_map! {
    "{" => "LeftCurly",
    "}" => "RightCurly",
    "(" => "LeftParen",
    ")" => "RightParen",
    "[" => "LeftBracket",
    "]" => "RightBracket",
    "," => "Comma",
    ":" => "Colon",
    ";" => "Semicolon",
    "." => "Dot",
    "'" => "Quote",
    "\"" => "DoubleQuote",
    "@" => "At",
    "!" => "Bang",
    "#" => "Hash",
    "$" => "Dollar",
    "%" => "Percent",
    "^" => "Caret",
    "&" => "Ampersand",
    "*" => "Star",
    "-" => "Minus",
    "_" => "Underscore",
    "+" => "Plus",
    "=" => "Equal",
    ">" => "Greater",
    "<" => "Less",
    "|" => "Pipe",
    "~" => "Tilde",
    "/" => "Slash",
    "\\" => "Backslash",
    "//" => "SlashSlash",
    "//=>" => "SlashSlashEqual",
    "//=" => "SlashSlashEqual",
    "/=" => "SlashEqual",
    "/>" => "SlashGreater",
    "/?" => "SlashQuestion",
    "/??" => "SlashNullish",
    "/*" => "SlashStar",
    "*/" => "StarSlash",
    "+++" => "PlusPlusPlus",
    "!!" => "BangBang",
    "!!=" => "BangBangEqual",
    "!!?" => "BangBangQuestion",
    "!!??" => "BangBangNullish",
    "!!???" => "BangBangNullishQuestion",
    "?" => "Question",
    "->" => "Arrow",
    "=>" => "FatArrow",
    "++" => "PlusPlus",
    "--" => "MinusMinus",
    "&&" => "And",
    "||" => "Or",
    "==" => "EqualEqual",
    "!=" => "NotEqual",
    ">=" => "GreaterEqual",
    "<=" => "LessEqual",
    "===" => "TripleEqual",
    "!==" => "NotTripleEqual",
    "<<" => "ShiftLeft",
    ">>" => "ShiftRight",
    ">>>" => "ShiftRightUnsigned",
    "+=" => "PlusEqual",
    "-=" => "MinusEqual",
    "*=" => "StarEqual",
    "%=" => "PercentEqual",
    "&=" => "AmpersandEqual",
    "|=" => "PipeEqual",
    "^=" => "CaretEqual",
    "&&=" => "AndEqual",
    "||=" => "OrEqual",
    "??=" => "NullishEqual",
    "??" => "Nullish",
    "???" => "NullishQuestion",
    "**" => "StarStar",
    "**=" => "StarStarEqual",
    "<>" => "LessGreater",
    "<=>" => "LessGreaterEqual",
    "<!" => "LessBang",
    "</" => "LessSlash",

};

pub static RUST_KEYWORDS: phf::Map<&'static str, &'static str> = phf::phf_map! {
    "abstract" => "Abstract",
    "as" => "As",
    "async" => "Async",
    "await" => "Await",
    "break" => "Break",
    "const" => "Const",
    "continue" => "Continue",
    "crate" => "Crate",
    "dyn" => "Dyn",
    "else" => "Else",
    "enum" => "Enum",
    "extern" => "Extern",
    "false" => "False",
    "final" => "Final",
    "fn" => "Fn",
    "for" => "For",
    "if" => "If",
    "impl" => "Impl",
    "in" => "In",
    "let" => "Let",
    "loop" => "Loop",
    "match" => "Match",
    "mod" => "Mod",
    "move" => "Move",
    "mut" => "Mut",
    "pub" => "Pub",
    "ref" => "Ref",
    "return" => "Return",
    "self" => "Self",
    "static" => "Static",
    "struct" => "Struct",
    "super" => "Super",
    "trait" => "Trait",
    "true" => "True",
    "type" => "Type",
    "unsafe" => "Unsafe",
    "use" => "Use",
    "where" => "Where",
    "while" => "While",
    "with" => "With",
    "yield" => "Yield",
    "None" => "_None",
    "Some" => "_Some",
    "Ok" => "_Ok",
    "Err" => "_Err",
    "Result" => "_Result",
    "Option" => "_Option",
    "Vec" => "_Vec",
};
