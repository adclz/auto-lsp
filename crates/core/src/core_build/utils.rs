/*
This file is part of auto-lsp.
Copyright (C) 2025 CLAUZEL Adrien

auto-lsp is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>
*/

pub(crate) fn intersecting_ranges(
    range1: &std::ops::Range<usize>,
    range2: &tree_sitter::Range,
) -> bool {
    range1.start <= range2.start_byte && range1.end >= range2.end_byte
}

pub(crate) fn tree_sitter_range_to_lsp_range(range: &tree_sitter::Range) -> lsp_types::Range {
    let start = range.start_point;
    let end = range.end_point;
    lsp_types::Range {
        start: lsp_types::Position {
            line: start.row as u32,
            character: start.column as u32,
        },
        end: lsp_types::Position {
            line: end.row as u32,
            character: end.column as u32,
        },
    }
}
