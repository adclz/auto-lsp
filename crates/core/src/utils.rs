#[macro_export]
macro_rules! dispatch_once {
    ($node:expr, [$($ty:ty => $method:ident($($param:expr),*)),*]) => {
        $(
            if let Some(n) = $node.downcast_ref::<$ty>() {
                return n.$method($($param),*);
            };
        )*
    };
}

#[macro_export]
macro_rules! dispatch {
    ($node:expr, [$($ty:ty => $method:ident($($param:expr),*)),*]) => {
        $(
            if let Some(n) = $node.downcast_ref::<$ty>() {
                n.$method($($param),*)?;
            };
        )*
    };
}