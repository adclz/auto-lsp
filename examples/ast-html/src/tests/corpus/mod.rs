/*
This file is part of auto-lsp.
Copyright (C) 2025 CLAUZEL Adrien

auto-lsp is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>
*/
use crate::{snap};
use utils::Result;
mod utils;

#[test]
fn tags() -> Result {
    snap!("<span>Hello</span>", tags)
}

#[test]
fn tags_with_attributes() -> Result {
    snap!(r#"<input value=yes class="a" data-üí©></input>"#, tags_with_attributes)
}


#[test]
fn nested_tags() -> Result {
    snap!(r#"<div>
  <span>a</span>
  b
  <b>c</b>
  Multi-line
  text
</div>"#, nested_tags)
}


#[test]
fn void_tags() -> Result {
    snap!(
        r#"<form><img src="something.png"><br><input type=submit value=Ok /></form>"#,
        void_tags
    )
}

#[test]
fn void_tags_at_eof() -> Result {
    snap!(
        r#"<img src="something.png">"#,
        void_tags_at_eof
    )
}

#[test]
fn custom_tags() -> Result {
    snap!(
        r#"<something:different>
  <atom-text-editor mini>
    Hello
  </atom-text-editor>
</something:different>"#,
        custom_tags
    )
}

#[test]
fn raw_text_elements() -> Result {
    snap!(
        r#"<script>
  </s
  </sc
  </scr
  </scri
  </scrip
</script>

<style>
  </ </s </st </sty </styl
</style>

<script>
</SCRIPT>"#,
        raw_text_elements
    )
}

#[test]
fn all_caps_doctype() -> Result {
    snap!(
        r#"<!DOCTYPE html PUBLIC
  "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">"#,
        all_caps_doctype
    )
}

#[test]
fn li_elements_without_close_tags() -> Result {
    snap!(
        r#"<ul>
  <li>One
  <li>Two
</ul>"#,
        li_elements_without_close_tags
    )
}

#[test]
fn dt_and_dl_elements_without_close_tags() -> Result {
    snap!(
        r#"<dl>
  <dt>Coffee
  <dt>Caf√©
  <dd>Black hot drink
  <dt>Milk
  <dd>White cold drink
</dl>"#,
        dt_and_dl_elements_without_close_tags
    )
}

#[test]
fn p_elements_without_close_tags() -> Result {
    snap!(
        r#"<p>One
<div>Two</div>
<p>Three
<p>Four
<h1>Five</h1>"#,
        p_elements_without_close_tags
    )
}

#[test]
fn ruby_annotation_elements_without_close_tags() -> Result {
    snap!(
        r#"<ruby>Êù±<rb>‰∫¨<rt>„Å®„ÅÜ<rt>„Åç„Çá„ÅÜ</ruby>"#,
        ruby_annotation_elements_without_close_tags
    )
}

#[test]
fn col_group_elements_without_end_tags() -> Result {
    snap!(
        r#"<table>
  <colgroup>
    <col style="background-color: #0f0">
    <col span="2">
  <tr>
    <th>Lime</th>
    <th>Lemon</th>
    <th>Orange</th>
  </tr>
</table>"#,
        col_group_elements_without_end_tags
    )
}

#[test]
fn tr_td_th_elements_without_end_tags() -> Result {
    snap!(
        r#"<table>
  <tr>
    <th>One
    <th>Two
  <tr>
    <td>Three
    <td>Four
</table>"#,
        tr_td_th_elements_without_end_tags
    )
}

#[test]
fn named_entities_in_tag_content() -> Result {
    snap!(
        r#"<p>Lorem ipsum &nbsp; dolor sit &copy; amet.</p>"#,
        named_entities_in_tag_content
    )
}

#[test]
fn numeric_entities_in_tag_content() -> Result {
    snap!(
        r#"<p>Lorem ipsum &#160; dolor sit &#8212; amet.</p>"#,
        numeric_entities_in_tag_content
    )
}

#[test]
fn multiple_entities_in_tag_content() -> Result {
    snap!(
        r#"<p>Lorem ipsum &#xA0; dolor &#xa0; sit &nbsp; amet.</p>"#,
        multiple_entities_in_tag_content
    )
}

#[test]
fn omitted_end_tags() -> Result {
    snap!(
        r#"<!doctype html><html><head>"#,
        omitted_end_tags
    )
}
