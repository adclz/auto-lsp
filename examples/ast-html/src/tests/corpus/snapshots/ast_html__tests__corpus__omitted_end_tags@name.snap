---
source: src/tests/corpus/mod.rs
expression: document
---
Document {
    children: [
        Doctype(
            Doctype {
                [RANGE]
                _id: 1,
                _parent: Some(
                    0,
                ),
            },
        ),
        Element(
            Element {
                children: [
                    StartTag(
                        StartTag {
                            children: [
                                TagName(
                                    TagName {
                                        [RANGE]
                                        _id: 4,
                                        _parent: Some(
                                            3,
                                        ),
                                    },
                                ),
                            ],
                            [RANGE]
                            _id: 3,
                            _parent: Some(
                                2,
                            ),
                        },
                    ),
                    Element(
                        Element {
                            children: [
                                StartTag(
                                    StartTag {
                                        children: [
                                            TagName(
                                                TagName {
                                                    [RANGE]
                                                    _id: 7,
                                                    _parent: Some(
                                                        6,
                                                    ),
                                                },
                                            ),
                                        ],
                                        [RANGE]
                                        _id: 6,
                                        _parent: Some(
                                            5,
                                        ),
                                    },
                                ),
                            ],
                            [RANGE]
                            _id: 5,
                            _parent: Some(
                                2,
                            ),
                        },
                    ),
                ],
                [RANGE]
                _id: 2,
                _parent: Some(
                    0,
                ),
            },
        ),
    ],
    [RANGE]
    _id: 0,
    _parent: None,
}
